internal/core/src/segcore/SegmentGrowingImpl.cpp

```diff
@@ -105,121 +106,84 @@ SegmentGrowingImpl::mask_with_delete(BitsetType& bitset, int64_t ins_barrier, Ti
+        auto field_id = FieldId(field.field_id());
+        AssertInfo(!field_id_to_offset.count(field_id), "duplicate field data");
+        field_id_to_offset.emplace(field_id, field_offset++);
     }
 
     // step 2: sort timestamp
-    auto raw_data = reinterpret_cast<const char*>(entities_raw.raw_data);
-    auto len_per_row = entities_raw.sizeof_per_row;
-    std::vector<std::tuple<Timestamp, idx_t, int64_t>> ordering;
-    ordering.resize(size);
-    // #pragma omp parallel for
-    for (int i = 0; i < size; ++i) {
-        ordering[i] = std::make_tuple(timestamps_raw[i], uids_raw[i], i);
-    }
-    std::sort(ordering.begin(), ordering.end());
+    // sort timestamps when flow graph receive insert msgs
+    // sort timestamps of internalPb.insertRequest in proxy
+    // avoid repeat copy in c++
 
     // step 3: and convert row-based data to column-based data accordingly
-    auto sizeof_infos = schema_->get_sizeof_infos();
-    std::vector<int> offset_infos(schema_->size() + 1, 0);
-    std::partial_sum(sizeof_infos.begin(), sizeof_infos.end(), offset_infos.begin() + 1);
-    std::vector<aligned_vector<uint8_t>> entities(schema_->size());
-
-    for (int fid = 0; fid < schema_->size(); ++fid) {
-        auto len = sizeof_infos[fid];
-        entities[fid].resize(len * size);
-    }
 
-    std::vector<idx_t> uids(size);
-    std::vector<Timestamp> timestamps(size);
-    // #pragma omp parallel for
-    for (int index = 0; index < size; ++index) {
-        auto [t, uid, order_index] = ordering[index];
-        timestamps[index] = t;
-        uids[index] = uid;
-        for (int fid = 0; fid < schema_->size(); ++fid) {
-            auto len = sizeof_infos[fid];
-            auto offset = offset_infos[fid];
-            auto src = raw_data + order_index * len_per_row + offset;
-            auto dst = entities[fid].data() + index * len;
-            memcpy(dst, src, len);
-        }
-    }
-
-    do_insert(reserved_begin, size, uids.data(), timestamps.data(), entities);
-    return Status::OK();
-}
-
-void
-SegmentGrowingImpl::do_insert(int64_t reserved_begin,
-                              int64_t size,
-                              const idx_t* row_ids,
-                              const Timestamp* timestamps,
-                              const std::vector<aligned_vector<uint8_t>>& columns_data) {
     // step 4: fill into Segment.ConcurrentVector
-    record_.timestamps_.set_data(reserved_begin, timestamps, size);
-    record_.uids_.set_data(reserved_begin, row_ids, size);
-    for (int fid = 0; fid < schema_->size(); ++fid) {
-        auto field_offset = FieldOffset(fid);
-        record_.get_field_data_base(field_offset)->set_data_raw(reserved_begin, columns_data[fid].data(), size);
+    insert_record_.timestamps_.set_data_raw(reserved_offset, timestamps_raw, size);
+    insert_record_.row_ids_.set_data_raw(reserved_offset, row_ids, size);
+    for (auto [field_id, field_meta] : schema_->get_fields()) {
```

Comment:
you can iterate "insert_data->fields_data()" instead of "schema_->get_fields()", 
in this way we need not generate "field_it_to_offset"

Reply:
field_id_to_offset is also used to check if there is a duplicate field_id in the inserted data